src/runtime-rs/crates/hypervisor/src/dragonball/inner.rs

```diff
@@ -0,0 +1,300 @@
+        if self.jailed {
+            self.umount_jail_resource(DRAGONBALL_KERNEL).ok();
+            self.umount_jail_resource(DRAGONBALL_ROOT_FS).ok();
+            for id in &self.cached_block_devices {
+                self.umount_jail_resource(id.as_str()).ok();
+            }
+        }
+
+        std::fs::remove_dir_all(&self.vm_path)
+            .map_err(|err| {
+                error!(sl!(), "failed to remove dir all for {}", &self.vm_path);
+                err
+            })
+            .ok();
+    }
+
+    fn set_vm_base_config(&mut self) -> Result<()> {
+        let serial_path = [&self.run_dir, "console.sock"].join("/");
+        let vm_config = VmConfigInfo {
+            serial_path: Some(serial_path),
+            mem_size_mib: self.config.memory_info.default_memory as usize,
+            vcpu_count: self.config.cpu_info.default_vcpus as u8,
+            ..Default::default()
+        };
+        info!(sl!(), "vm config: {:?}", vm_config);
+
+        self.vmm_instance
+            .set_vm_configuration(vm_config)
+            .context("set vm configuration")
+    }
+
+    pub(crate) fn umount_jail_resource(&self, jailed_path: &str) -> Result<()> {
+        let path = [self.jailer_root.as_str(), jailed_path].join("/");
+        nix::mount::umount2(path.as_str(), nix::mount::MntFlags::MNT_DETACH)
+            .with_context(|| format!("umount path {}", &path))
+    }
+
+    pub(crate) fn get_resource(&self, src: &str, dst: &str) -> Result<String> {
+        if self.jailed {
+            self.jail_resource(src, dst)
+        } else {
+            Ok(src.to_string())
+        }
+    }
+
+    fn jail_resource(&self, src: &str, dst: &str) -> Result<String> {
+        info!(sl!(), "src {} dst {}", src, dst);
+        if src.is_empty() || dst.is_empty() {
+            return Err(anyhow!("invalid param src {} dst {}", src, dst));
+        }
+
+        let jailed_location = [self.jailer_root.as_str(), dst].join("/");
+        mount::bind_mount_unchecked(src, jailed_location.as_str(), false).context("bind_mount")?;
+
+        let mut abs_path = String::from("/");
+        abs_path.push_str(dst);
+        Ok(abs_path)
+    }
+
+    fn set_boot_source(&mut self, kernel_path: &str, kernel_params: &str) -> Result<()> {
+        info!(
+            sl!(),
+            "kernel path {} kernel params {}", kernel_path, kernel_params
+        );
+
+        let mut boot_cfg = BootSourceConfig {
+            kernel_path: self
+                .get_resource(kernel_path, DRAGONBALL_KERNEL)
+                .context("jail resource")?,
+            ..Default::default()
+        };
+
+        if !kernel_params.is_empty() {
+            boot_cfg.boot_args = Some(kernel_params.to_string());
+        }
+
+        self.vmm_instance
+            .put_boot_source(boot_cfg)
+            .context("put boot source")
+    }
+
+    fn set_vm_rootfs(&mut self, path: &str, driver: &str) -> Result<()> {
+        info!(sl!(), "==== set vm rootfs {} {}", path, driver);
```

Comment:
Nit: remove `====`

Reply:
Thanks for your review, already remove the "====".